src/backend/distributed/executor/adaptive_executor.c

```diff
@@ -4080,6 +4086,18 @@ WorkerSessionFailed(WorkerSession *session)
 
 		PlacementExecutionDone(placementExecution, succeeded);
 	}
+
+	/*
+	 * We might hit the connection timeout. In that case, we should not
+	 * let the connection establishment to continue because the execution
+	 * logic pretends that failed sessions are not going to be used anymore.
+	 */
+	MultiConnection *connection = session->connection;
+	if (connection->connectionState == MULTI_CONNECTION_INITIAL ||
+		connection->connectionState == MULTI_CONNECTION_CONNECTING)
+	{
+		connection->connectionState = MULTI_CONNECTION_FAILED;
+	}
```

Comment:
it's a little odd that CitusPQFinish does not set any state. Would that be worth changing? (I'm not sure to what)

Reply:
We should probably have a new state as none of the states cover what we need.

It might be nicer to do it with a separate PR?

Reply:
if we set it to something like MULTI_CONNECTION_CLOSED, do we still need this if statement?

Reply:
So, you mean we call `CitusPQFinish()` here and that already triggers `ConnectionStateMachine`'s connection failure step?

That could potentially work, let me think/test a bit further.

Reply:
That could work but it makes the distinction between `CONNECTION_FAILED` vs `CONNECTION_LOST` confusing. Also, it makes this change invasive which becomes harder to backport?

What would  be your take to the latest commit?

Reply:
I was thinking about the ShutdownConnection->CitusPQFinish call in ConnectionStateMachine, but I guess part of the problem is that we don't call that for all connections?

Reply:
Yes, that's right. The connections that timeout happened might still try to establish the connection and succeed. 

With this code, it'd be as-if `PQstatus` returns `CONNECTION_BAD` while polling and the execution cleans up the relevant states. 

One another option I considered was to call `ShutdownConnection()` right here so that `ConnectionStateMachine()` would eventually  get to `MULTI_CONNECTION_FAILED`. One concern is that  cancellation code in `ShutdownConnection()`